subprojects/language-jvm/src/main/java/org/gradle/api/tasks/compile/ProviderAwareForkOptions.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.tasks.compile;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Lists;
+import org.gradle.api.Incubating;
+import org.gradle.api.tasks.Internal;
+import org.gradle.api.tasks.Nested;
+import org.gradle.api.tasks.Optional;
+import org.gradle.process.CommandLineArgumentProvider;
+import org.gradle.util.internal.CollectionUtils;
+
+import java.util.List;
+
+/**
+ * Fork options for compilation that can accept user-defined {@link CommandLineArgumentProvider} objects.
+ *
+ * Only take effect if {@code fork} is {@code true}.
+ *
+ * @since 7.1
+ */
+@Incubating
+public class ProviderAwareForkOptions extends BaseForkOptions {
+
+    private final List<CommandLineArgumentProvider> jvmArgumentProviders = Lists.newArrayList();
+
+    /**
+     * Returns any additional JVM argument providers for the compiler process.
+     *
+     */
+    @Optional
+    @Nested
+    public List<CommandLineArgumentProvider> getJvmArgumentProviders() {
+        return jvmArgumentProviders;
+    }
+
+    @Internal
+    public List<String> getAllJvmArgs() {
+        ImmutableList.Builder<String> builder = ImmutableList.builder();
+        builder.addAll(CollectionUtils.stringize(getJvmArgs()));
+        for (CommandLineArgumentProvider argumentProvider : getJvmArgumentProviders()) {
+            builder.addAll(argumentProvider.asArguments());
+        }
+        return builder.build();
+    }
+
+    @Override
+    protected boolean excludeFromAntProperties(String fieldName) {
```

Comment:
Do you know if we still need to exclude things from Ant properties here and what this is for?

Reply:
I don't think we do, actually.  I suspect this is vestigial cruft from back when there was an option to compile using the ant task.  That was pulled out ages ago, but I think this has lingered.  `AbstractOptions` is the parent of a number of other unrelated objects, though, so pulling this out could have a broader impact.  Not to mention there might be issues with pulling it out that I'm not yet aware of.  I suggest we pull this out as a secondary follow up PR rather than lumping it in with this one.

Reply:
Yeah, let's do this in a follow up.